packages/core/capabilities/core-capabilities-server/src/contracts.ts

```diff
@@ -74,44 +74,81 @@ export interface CapabilitiesSetup {
    *          }
    *        }
    *      }
-   *      if(myPluginApi.shouldRestrictSomePluginBecauseOf(request)) {
+   *      if(await myPluginApi.shouldRestrictSomePluginBecauseOf(request)) {
    *        return {
    *          somePlugin: {
    *            featureEnabledByDefault: false // `featureEnabledByDefault` will be disabled. All other capabilities will remain unchanged.
    *          }
    *        }
    *      }
    *      return {}; // All capabilities will remain unchanged.
+   *    }, {
+   *      // the switcher only toggles capabilities under the 'somePlugin' path
+   *      capabilityPath: 'somePlugin',
    *    });
    * }
    * ```
    */
-  registerSwitcher(switcher: CapabilitiesSwitcher): void;
+  registerSwitcher(switcher: CapabilitiesSwitcher, options: CapabilitiesSwitcherOptions): void;
 }
 
 /**
- * Defines a set of additional options for the `resolveCapabilities` method of {@link CapabilitiesStart}.
+ * APIs to access the application {@link Capabilities}.
+ *
+ * @public
+ */
+export interface CapabilitiesStart {
+  /**
+   * Resolve the {@link Capabilities} to be used for given request
+   *
+   * @param request The request to resolve capabilities for
+   * @param options.capabilityPath The path(s) of the capabilities that needs to be retrieved. Use '*' to retrieve all paths.
+   * Used to avoid unnecessarily running switched on parts of the capabilities that won't be used by the API consumer.
+   *
+   * @example
+   * ```ts
+   * const mlCapabilities = (await coreStart.capabilities.resolveCapabilities(request, 'ml')).ml;
+   * ```
+   */
+  resolveCapabilities(
+    request: KibanaRequest,
+    options: ResolveCapabilitiesOptions
+  ): Promise<Capabilities>;
+}
+
+/**
+ * Options for {@link CapabilitiesStart.resolveCapabilities}.
  *
  * @public
  */
 export interface ResolveCapabilitiesOptions {
+  /**
+   * The path(s) of capabilities that the API consumer is interested in. The '*' wildcard is supported as a suffix only.
+   *
+   * E.g. capabilityPath: "*" or capabilityPath: "myPlugin.*" or capabilityPath: "myPlugin.myKey"
+   *
+   * @remark All the capabilities will be returned, but the ones not matching the specified path(s) may not have been processed
+   *        by the capability switchers and should not be used.
+   */
+  capabilityPath: string | string[];
```

Comment:
I'd suggest we create a `type CapabilityPath` type that can be used here and on line 153 below.  That way we need to update a single type in future.